pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,270 @@
+			if err != nil {
+				err = fmt.Errorf("failed to retrieve new table ID: %w", err)
+				errors = append(errors, err)
+				scopedLog.WithError(err).
+					Debug("Skipping migration of egress rule due to failure in retrieving new table ID")
+				continue
+			}
+
+			if err := migrateOnUpgrade(oldTable, newTable); err != nil {
+				err = fmt.Errorf("failed to migrate ENI rules on upgrade: %w", err)
+				errors = append(errors, err)
+				scopedLog.WithError(err).
+					Debug("Skipping migration of egress rule due to failure in migration")
+				continue
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	return errors
+}
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. If table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex.
+func retrieveTableIDFromIfIndex(ifindex int, getter interfaceDB) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, err
+	}
+
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(link.Attrs().HardwareAddr.String())
+	if err != nil {
+		return -1, err
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := copyRoutes(old, new); err != nil {
+		return err
+	}
+	if err := createNewRules(
+		linux_defaults.RulePriorityEgress,
+		old,
+		linux_defaults.RulePriorityEgressv2,
+		new,
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress, old); err != nil {
+		return err
+	}
+	if err := deleteOldRoutes(old); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyRoutes(from, to int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: from,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return err
+	}
+
+	// In ENI mode, we only expect two rules:
+	//  1) Link scoped route with a gateway IP
+	//  2) Default route via gateway IP
+	// We must upsert the link scoped route first which will make the kernel
+	// aware of the gateway IP. After that, then we will upsert the "default
+	// via gatewayIP" route. If we didn't do this, then the kernel would
+	// complain with "Error: Nexthop has invalid gateway." with an errno of
+	// ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := netlink.RouteReplace(&route); err != nil {
+				return fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			break
+		}
+	}
+
+	for _, r := range routes {
+		route := r
+		route.Table = to
```

Comment:
You probably want to skip `r.Scope == netlink.SCOPE_LINK` to not insert the routes already inserted a second time.